crates/sui-types/src/temporary_store.rs

```diff
@@ -108,48 +118,248 @@ impl<S> TemporaryStore<S> {
+            };
+
+        for (id, (ctx, obj, kind)) in self.written {
+            let old_obj = self.input_objects.get(&id);
+            let written_events =
+                Self::create_written_events(ctx, kind, id, &obj, old_obj, gas_id, gas_charged);
+            events.extend(written_events);
+            written.insert(id, (obj.compute_object_reference(), obj, kind));
+        }
+
+        for (id, (ctx, version, kind)) in self.deleted {
+            // Create events for each deleted changes
+            let deleted_obj = self.input_objects.get(&id);
+            let balance = deleted_obj
+                .and_then(|o| Coin::extract_balance_if_coin(o).ok())
+                .flatten();
+
+            let event = match (deleted_obj, balance) {
+                // Object is an owned (provided as input) coin object, create a spend event for the remaining balance.
+                (Some(deleted_obj), Some(balance)) => {
+                    let balance = balance as i128;
+                    Event::balance_change(
+                        &ctx,
+                        BalanceChangeType::Pay,
+                        deleted_obj.owner,
+                        id,
+                        deleted_obj.version(),
+                        deleted_obj.type_().unwrap(),
+                        balance.neg(),
+                    )
+                }
+                // If deleted object is not owned coin coin, emit a delete event.
+                _ => Event::DeleteObject {
+                    package_id: ctx.package_id,
+                    transaction_module: ctx.transaction_module.clone(),
+                    sender: ctx.sender,
+                    object_id: id,
+                    version,
+                },
+            };
+            events.push(event);
+            deleted.insert(id, (version, kind));
+        }
+
+        // Combine object events with move events.
+        events.extend(self.events);
+
+        let store = InnerTemporaryStore {
+            objects: self.input_objects,
+            mutable_inputs: self.mutable_input_refs,
+            written,
+            deleted,
+        };
+        (store, events)
+    }
+
+    fn create_written_events(
+        ctx: SingleTxContext,
+        kind: WriteKind,
+        id: ObjectID,
+        obj: &Object,
+        old_obj: Option<&Object>,
+        gas_id: Option<ObjectID>,
+        gas_charged: i128,
+    ) -> Vec<Event> {
+        match (kind, Coin::extract_balance_if_coin(obj), old_obj) {
+            // For mutation of existing coin, we need to compute the coin balance delta
+            // and emit appropriate event depends on ownership changes
+            (WriteKind::Mutate, Ok(Some(_)), Some(old_obj)) => {
+                Self::create_coin_mutate_events(&ctx, gas_id, obj, old_obj, gas_charged)
+            }
+            // For all other coin change (wrap/unwrap/create), we emit full balance transfer event to the new owner.
+            (_, Ok(Some(balance)), _) => {
```

Comment:
Wrap is a DeleteKind not a WriteKind? 
I think it makes sense to count unwrap & create as balance_change as they would result balance change indeed, while it might also be helpful to emit create & unwrap events as well, for example ppl might want to count on these events to track their coin objects.

Reply:
balanceChange event include the coin object id as well, so user can kind of use this information to work out if this is a new coin or not.... trying to not create too many new event at this stage as this will increase work load to the event store (which might become a bottle neck) ...

Reply:
it is a trade-off indeed, emitting those sorta duplicate events would make it easier and less err prone on the consumer side though, for example as the other comment mentioned when ppl want to track some token objects counting on certain type of events. 
def. not blocking here but if the headroom turns out okay for us to have some extra events, I think it's worth it to simplify the client side logic if we can do that.